libs/etcd_broker/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+//! A set of primitives to access a shared data/updates, propagated via etcd broker (not persistent).
+//! Intended to connect services to each other, not to store their data.
+use std::{
+    collections::{hash_map, HashMap},
+    str::FromStr,
+};
+
+use anyhow::Context;
+use regex::{Captures, Regex};
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DisplayFromStr};
+
+pub use etcd_client::*;
+
+use tokio::{sync::mpsc, task::JoinHandle};
+use tracing::*;
+use utils::{
+    lsn::Lsn,
+    zid::{ZNodeId, ZTenantId, ZTenantTimelineId},
+};
+
+#[derive(Debug, Deserialize, Serialize)]
+struct SafekeeperTimeline {
+    safekeeper_id: ZNodeId,
+    info: SkTimelineInfo,
+}
+
+/// Published data about safekeeper's timeline. Fields made optional for easy migrations.
+#[serde_as]
+#[derive(Debug, Deserialize, Serialize)]
+pub struct SkTimelineInfo {
+    /// Term of the last entry.
+    pub last_log_term: Option<u64>,
+    /// LSN of the last record.
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub flush_lsn: Option<Lsn>,
+    /// Up to which LSN safekeeper regards its WAL as committed.
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub commit_lsn: Option<Lsn>,
+    /// LSN up to which safekeeper offloaded WAL to s3.
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub s3_wal_lsn: Option<Lsn>,
+    /// LSN of last checkpoint uploaded by pageserver.
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub remote_consistent_lsn: Option<Lsn>,
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub peer_horizon_lsn: Option<Lsn>,
+    #[serde(default)]
+    pub wal_stream_connection_string: Option<String>,
+}
+
+// TODO: add global zenith installation ID.
+const ZENITH_BROKER_PREFIX: &str = "zenith";
```

Comment:
Some time ago we had a discussion that we're using single etcd for several neon installations. So I think we can take this const from `ETCD_BROKER_PREFIX` envvar. It's not urgent, we can create an issue and do it later.

Reply:
Will adjust, thanks. Also have to change it to `neon` anyway..

Reply:
Have thought about it more and placed that into `SafekeeperConf` without using env vars at all.
Lmk what you think.

Reply:
Yeah, thought of it too, but forgot to mention. We currently have different approaches to configuration in safekeeper and pageserver, but we need this prefix to be the same for both. `SafekeeperConf` is ok, but I guess we should also add it to pageserver config then.